### PR TITLE
Adding a `ClientIdentityFromSPIFFE()` helper.

### PR DESCRIPTION
**Pull request by dhermes**
## PR Summary

 - **Type:** Features
 - **Intended Change Level:** minor

#### Reviewers:

Reviewers keep the following in mind while reviewing this PR, and provide an assessment of them in your approval comment:

- Does the "Intended Change Level" above match the actual behavior of this PR?
- Is this PR following patterns set forth in the package (if modifying a package), or the go-sdk design patterns if implementing a new package from scratch?
- Does this require a backport to LTS versions? If so ping, let the contributors group know.

### CODE REVIEW

File: envoyutil/identity.go [hunk 1]
```diff
@@ -68,3 +75,64 @@ func ExtractAndVerifyClientIdentity(req *http.Request, cip ClientIdentityProvide
...
+// ClientIdentityFromSPIFFE produces a function satisfying `ClientIdentityProvider`.
+//
+// This function assumes the client identity is in the `URI` field and that field
+// is a SPIFFE URI. Further, it assumes the workload identifier identifies a
+// Kubernetes service account, of the form  `ns/{namespace}/sa/{serviceAccount}`.
+// After parsing, the returned value is of the form `{namespace}.{serviceAccount}`.
+//
+// Additionally, it takes a variadic input of `denied` client identities that
+// should not pass validation.
+func ClientIdentityFromSPIFFE(trustDomain string, denied ...string) ClientIdentityProvider {
```
Comment (dhermes) [inline envoyutil/identity.go:112]: @wcharczuk I am a bit torn on whether `denied` should be included in `ClientIdentityFromSPIFFE()` or if there should be two functions exposed. Open to suggestions.

File: envoyutil/identity.go [hunk 2]
```diff
@@ -68,3 +75,64 @@ func ExtractAndVerifyClientIdentity(req *http.Request, cip ClientIdentityProvide
...
+	// NOTE: The `pu == nil` check is redundant, we expect `spiffeutil.Parse()`
+	//       not to violate the invariant that `pu != nil` when `err == nil`.
+	if err != nil || pu == nil {
+		return "", &XFCCExtractionError{
+			Class: ErrInvalidClientIdentity,
+			XFCC:  xfcc.String(),
+		}
+	}
+
+	if pu.TrustDomain != trustDomain {
```
Comment (wcharczuk) [inline envoyutil/identity.go:93]: are there any casing issues here? do we want to use `strings.EqualFold`?

File: envoyutil/identity.go [hunk 3]
```diff
@@ -68,3 +75,64 @@ func ExtractAndVerifyClientIdentity(req *http.Request, cip ClientIdentityProvide
 	// Do final extraction.
 	return cip(xfcc)
 }
+
+// ClientWorkloadID determines the SPIFFE workload identifier for the client.
+// This function assumes the client identity is in the `URI` field and that field
+// is a SPIFFE URI.
+func ClientWorkloadID(trustDomain string, xfcc XFCCElement) (string, error) {
```
Comment (wcharczuk) [inline envoyutil/identity.go:82]: I would add something in the comments to the effect of that the trust domain will be validated.

Comment (wcharczuk) [inline envoyutil/identity.go:82]: another comment here is this function is a bit more involved than the noun `ClientWorkloadID` would indicate.

I would have this function take an options list that mutates a `WorkloadProcessor` or similar, where you can add `AllowList` domains, `DenyList` domains, and potentially other custom validators, a method on this type would then be returned as the delegate return.

File: envoyutil/client_identity_processor.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+// ClientIdentityFormatter describes functions that will produce a client
+// identity string from a parsed SPIFFE URI.
+type ClientIdentityFormatter = func(XFCCElement, *spiffeutil.ParsedURI) (string, error)
+
+// ClientIdentityProcessor provides configurable fields that can be used to
+// help validate a parsed SPIFFE URI and produce and validate a client identity
+// from a parsed SPIFFE URI.
+type ClientIdentityProcessor struct {
+	AllowedTrustDomains     []string
+	DeniedTrustDomains      []string
```
Comment (dhermes) [inline envoyutil/client_identity_processor.go:83]: @wcharczuk When you introduced `AllowedTrustDomains` and `DeniedTrustDomains` I wonder if you were mixing up my usage of

```go
if pu.TrustDomain != trustDomain {
	// ...
}
```

and

```go
if deniedSet.Contains(clientIdentity) {
	// ...
}
```

(See original commit: https://github.com/blend/go-sdk/pull/208/commits/e75963212222cf7844b03b12c207e9086b42fabc)

Based on the usage of

```go
		cip := envoyutil.ClientIdentityFromSPIFFE(
			envoyutil.OptAllowedTrustDomains(tc.TrustDomain),
			envoyutil.OptDeniedTrustDomains(tc.Denied...),
		)
```

in the unit tests in https://github.com/blend/go-sdk/pull/209, it seems you had conflated the two concepts by understandable review mistake.

----

IMO having `AllowedTrustDomains` and `DeniedTrustDomains` is likely not necessary. Pinning to a single trust domain for a k8s cluster (i.e. `cluster.local`) and then doing case-insensitive matching on it should be sufficient.

Having 2 allow lists and 2 deny lists is a bit of overhead for this type, but if you think it's worth it, we can leave them as-is.

File: envoyutil/client_identity_processor.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+		Metadata: map[string]string{
+			"clientID": clientID,
+		},
+	}
+}
+
+// ProcessDeniedClientIdentities returns an error if a denied list is configured
+// and a client ID matches any elements in the list.
+func (cip ClientIdentityProcessor) ProcessDeniedClientIdentities(xfcc XFCCElement, clientID string) error {
+	if cip.DeniedClientIdentities.Len() == 0 {
```
Comment (dhermes) [inline envoyutil/client_identity_processor.go:186]: Not sure if it's worth it to call `.Len()` vs. just using `len(cip.DeniedClientIdentities)`. Ditto for `len(cip.AllowedClientIdentities)`.

File: envoyutil/xfcc.go [hunk 6]
```diff
@@ -236,7 +236,7 @@ func ParseXFCC(header string) (XFCC, error) {
 		}
 
 		// Increment the index for the next iteration. (Note that branches of the
-		// `switch` statement may have already incremented `i` as well.)
+		// `switch` statement may have already incremented `xp.Index` as well.)
```
Comment (dhermes) [inline envoyutil/xfcc.go:239]: I missed this in #212.

**Comment by dhermes** [comment]
Closed in favor of #213.

**Pull request closed without merging**